File: mobile_client/app/src/main/java/wpam/mobile_client/protocol/PlotResponse.java
package wpam.mobile_client.protocol;

import java.util.ArrayList;

import wpam.mobile_client.sensor_tag.ParameterType;

public final class PlotResponse extends Response
{
    private ArrayList<Integer> values;
    private ArrayList<Integer> timestamps;
    private ParameterType parameterType;

    public PlotResponse(ArrayList<Integer> values, ArrayList<Integer> timestamps, ParameterType parameterType) {
        super(10, ResponseType.PLOT);
        this.values = values;
        this.timestamps = timestamps;
        this.parameterType = parameterType;
    }
    public final void close()
    {
        super.close();
    }

    @Override
    public ArrayList<Integer> getFrameBytes()
    {
        initializeDataSize();

        ArrayList<Integer> frame = getHeader();
        frame.add(responseType_.getValue());
        frame.add(parameterType.getValue());

        for ( Integer value: values) {
            frame.add(value);
        }

        for ( Integer time: timestamps) {
            frame.add(time);
        }
        return frame;
    }

    @Override
    public String getName()
    {
        return new String("BlindsStatusResponse");
    }

    @Override
    public void accept(ResponseVisitor visitor)
    {
        visitor.visit(this);
    }

    public ArrayList<Integer> getValues()
    {
        return this.values;
    }

    public void setValues(ArrayList<Integer> values)
    {
        this.values = values;
    }

    public ArrayList<Integer> getTimestamps() {
        return timestamps;
    }

    public void setTimestamps(ArrayList<Integer> timestamps) {
        this.timestamps = timestamps;
    }

    public ParameterType getParameterType() {
        return parameterType;
    }

    public void setParameterType(ParameterType parameterType) {
        this.parameterType = parameterType;
    }

    @Override
    protected void initializeDataSize()
    {
        setDataSize(10);
    }
}

File: server/src/protocol/PlotResponse.cpp
#include "PlotResponse.h"
#include "ResponseVisitor.h"
#include "CommandVisitor.h"

#include <utility/BytesConverter.h>

using namespace std;
using namespace utility;
using namespace communication;

PlotResponse::PlotResponse(const std::vector<uint16_t> &data, const std::vector<uint16_t> &timestamp,
                           ParameterType parameterType)
        : Response(10, ResponseType::PLOT_RESPONSE),
          data_(data),
          timestamp_(timestamp),
          parameterType_(parameterType)
{}

PlotResponse::~PlotResponse()
{}

std::vector<uint8_t> PlotResponse::getFrameBytes()
{
    initializeDataSize();

    vector<uint8_t > frame = getHeader();
    frame.push_back(static_cast<uint8_t>(responseType_));
    frame.push_back(static_cast<uint8_t>(parameterType_));

    for(auto sample : data_) {
        BytesConverter::appendUINT16toVectorOfUINT8(sample, frame);
    }

    for(auto time : timestamp_) {
        BytesConverter::appendUINT16toVectorOfUINT8(time, frame);
    }

    return frame;
}

std::string PlotResponse::getName()
{
    return string("PlotCommand");
}

void PlotResponse::accept(ResponseVisitor &visitor)
{
    visitor.visit(*this);
}

void PlotResponse::initializeDataSize()
{
    uint16_t dataSize = sizeof(responseType_);
    dataSize += sizeof(parameterType_);
    dataSize += sizeof(uint16_t)*data_.size();
    dataSize += sizeof(uint16_t)*timestamp_.size();

    setDataSize(dataSize);
}

const std::vector<uint16_t> PlotResponse::getData() const
{
    return data_;
}

void PlotResponse::setData(const std::vector<uint16_t> &data)
{
    data_ = data;
}
const vector<uint16_t> &PlotResponse::getTimestamp() const
{
    return timestamp_;
}

void PlotResponse::setTimestamp(const vector<uint16_t> &timestamp)
{
    timestamp_ = timestamp;
}

ParameterType PlotResponse::getParameterType() const
{
    return parameterType_;
}

void PlotResponse::setParameterType(ParameterType parameterType)
{
    parameterType_ = parameterType;
}

